single-registry-architecture/README.md

```diff
@@ -0,0 +1,217 @@
+    like exactly what w need.
+  - Tested, I unfortunately [couldn't make it work](https://github.com/mcronce/oci-registry/issues/14) even for a simple
+  - test case, doesn't seem to be well supported.
+- [ACR connected registry](https://learn.microsoft.com/en-us/azure/container-registry/intro-connected-registry)
+        - there's no mention of how to deploy outside of AKS edge cluster, seems Azure IoT edge thing only
+- [docker's distribution/distribution](https://github.com/distribution/distribution)
+  - one instance can proxy only for a single upstream registry (but that's OK for us)
+  - tested, works with `containerd` for single upstream repo, works as well with `helm` charts
+  - no extra dependencies, can work with just local filesystem storage
+  - exposes reasonable prometheus metrics (transfer times, cache hit ratio)
+  - <https://github.com/distribution/distribution/blob/main/docs/configuration.md#proxy>
+  - <https://github.com/docker/docs/blob/main/content/registry/recipes/mirror.md>
+
+As a result, it seems we can use the `distribution` project from docker.
+
+## Implementation notes
+
+Very short and brief implementation idea. We can plan more details once we agree on the idea and solution.
+To keep the migration simpler, we split it separately for the container images and for the helm charts cache.
+We want to start with the images cache, as that's what we currently use Docker Hub for.
+
+### Container images
+
+Currently, we use two ACR registries, `giantswarm` for container images and `giantswarmpublic` for helm charts.
+The images one is already over 4 TB in size and it will be hard to clean it up. We can instead start with a new
+empty repo and populate it using a separate instance of `retagger`. We won't start uploading build artifacts there
+before we figure out how to use the retention policy.
+
+High level migration plan:
+
+- bootstrap a new ACR registry for images, let's call it tentatively `gsoci`
+  - configure access, geo-replication and a retention policy for the repo
+- prepare and deploy `distribution` cache instance, configure it for the new repo
+  - ensure monitoring and alerting
+- switch our CI/CD pipelines to not tag dev build artifacts
+  - use a tool (`skopeo`, `retagger`?) to replicate all (do we really need all?) the tagged versions of images we build from the old `giantswarm` repo to the new `gsoci`
+  - configure CI/CD so it uploads to both the old `giantswarm` and the new `gsoci`
+- deploy a new instance of `retagger` that will replicate all the configured public images to the new `gsoci` registry
+- switch container runtimes to use the cache as a source of images and upstream `gsoci` as a fallback
+- cleanup (after some time)
+  - delete our registries in docker hub and quay
+  - delete the old `giantswarm` registry in ACR
+- for this part, we keep China registry unchanged
+  - we reconfigure the `crsync` to replicate images from `gsoci` to `aliyun` in China
```

Comment:
My top of the head thought on this is the delay we introduce using crsync for this instead of uploading to a "primary" registry from the build.

Reply:
Sure, that's another approach. But, remember all those failed builds or builds that take 30 mins instead of 5 mins, where the only reason for that is that upload to China is slow or fails? We can only pick our poison here...

Reply:
I added a note that we might do one or the other. To get started, we won't change anything for China anyway.